pkg/kubelet/kubelet_stats.go

```diff
@@ -0,0 +1,644 @@
+	if rootFsInfo.Inodes != nil && rootFsInfo.InodesFree != nil {
+		nodeFsIU := *rootFsInfo.Inodes - *rootFsInfo.InodesFree
+		nodeFsInodesUsed = &nodeFsIU
+	}
+
+	// Get the root container stats's timestamp, which will be used as the
+	// imageFs stats timestamp.
+	rootStats, err := getCgroupStats(kl.cadvisor, "/")
+	if err != nil {
+		return nil, fmt.Errorf("failed to get root container stats: %v", err)
+	}
+
+	return &statsapi.FsStats{
+		Time:           metav1.NewTime(rootStats.Timestamp),
+		AvailableBytes: &rootFsInfo.Available,
+		CapacityBytes:  &rootFsInfo.Capacity,
+		UsedBytes:      &rootFsInfo.Usage,
+		InodesFree:     rootFsInfo.InodesFree,
+		Inodes:         rootFsInfo.Inodes,
+		InodesUsed:     nodeFsInodesUsed,
+	}, nil
+}
+
+// GetContainerInfo returns stats (from cAdvisor) for a container.
+func (kl *Kubelet) GetContainerInfo(podFullName string, podUID types.UID, containerName string, req *cadvisorapiv1.ContainerInfoRequest) (*cadvisorapiv1.ContainerInfo, error) {
+	// Resolve and type convert back again.
+	// We need the static pod UID but the kubecontainer API works with types.UID.
+	podUID = types.UID(kl.podManager.TranslatePodUID(podUID))
+
+	pods, err := kl.runtimeCache.GetPods()
+	if err != nil {
+		return nil, err
+	}
+	pod := kubecontainer.Pods(pods).FindPod(podFullName, podUID)
+	container := pod.FindContainerByName(containerName)
+	if container == nil {
+		return nil, kubecontainer.ErrContainerNotFound
+	}
+
+	ci, err := kl.cadvisor.DockerContainer(container.ID.ID, req)
+	if err != nil {
+		return nil, err
+	}
+	return &ci, nil
+}
+
+// GetRawContainerInfo returns the stats (from cadvisor) for a non-Kubernetes
+// container.
+func (kl *Kubelet) GetRawContainerInfo(containerName string, req *cadvisorapiv1.ContainerInfoRequest, subcontainers bool) (map[string]*cadvisorapiv1.ContainerInfo, error) {
+	if subcontainers {
+		return kl.cadvisor.SubcontainerInfo(containerName, req)
+	} else {
+		containerInfo, err := kl.cadvisor.ContainerInfo(containerName, req)
+		if err != nil {
+			return nil, err
+		}
+		return map[string]*cadvisorapiv1.ContainerInfo{
+			containerInfo.Name: containerInfo,
+		}, nil
+	}
+}
+
+// GetVersionInfo returns information about the version of cAdvisor in use.
+func (kl *Kubelet) GetVersionInfo() (*cadvisorapiv1.VersionInfo, error) {
+	return kl.cadvisor.VersionInfo()
+}
+
+// GetCachedMachineInfo assumes that the machine info can't change without a reboot
+func (kl *Kubelet) GetCachedMachineInfo() (*cadvisorapiv1.MachineInfo, error) {
+	if kl.machineInfo == nil {
+		info, err := kl.cadvisor.MachineInfo()
+		if err != nil {
+			return nil, err
+		}
+		kl.machineInfo = info
+	}
+	return kl.machineInfo, nil
+}
+
+// cadvisorInfoToContainerStats returns the statsapi.ContainerStats converted
+// from the container and filesystem info.
+func cadvisorInfoToContainerStats(name string, info *cadvisorapiv2.ContainerInfo, rootFs, imageFs *cadvisorapiv2.FsInfo) *statsapi.ContainerStats {
```

Comment:
For my own sanity, this is from the old `containerInfoV2ToStats` and `containerInfoV2FsStats`
Do we have any tests for these functions?